src/ImageSharp/Common/Helpers/DenseMatrixUtils.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Copyright (c) Six Labors and contributors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Numerics;
+using SixLabors.ImageSharp.Memory;
+using SixLabors.ImageSharp.PixelFormats;
+using SixLabors.ImageSharp.Primitives;
+
+namespace SixLabors.ImageSharp
+{
+    /// <summary>
+    /// Extension methods for <see cref="DenseMatrix{T}"/>.
+    /// </summary>
+    internal static class DenseMatrixUtils
+    {
+        /// <summary>
+        /// Computes the sum of vectors in <paramref name="targetRow"/> weighted by the kernel weight values.
+        /// </summary>
+        /// <typeparam name="TPixel">The pixel format.</typeparam>
+        /// <param name="matrix">The dense matrix.</param>
+        /// <param name="sourcePixels">The source frame.</param>
+        /// <param name="targetRow">The target row.</param>
+        /// <param name="row">The current row.</param>
+        /// <param name="column">The current column.</param>
+        /// <param name="maxRow">The maximum working area row.</param>
+        /// <param name="maxColumn">The maximum working area column.</param>
+        /// <param name="offsetColumn">The column offset to apply to source sampling.</param>
+        public static void Convolve<TPixel>(
+            in DenseMatrix<float> matrix,
+            Buffer2D<TPixel> sourcePixels,
+            Span<Vector4> targetRow,
+            int row,
+            int column,
+            int maxRow,
+            int maxColumn,
+            int offsetColumn)
+            where TPixel : struct, IPixel<TPixel>
+        {
+            Vector4 vector = default;
+            int matrixHeight = matrix.Rows;
+            int matrixWidth = matrix.Columns;
+            int radiusY = matrixHeight >> 1;
+            int radiusX = matrixWidth >> 1;
+
+            for (int y = 0; y < matrixHeight; y++)
+            {
+                int offsetY = (row + y - radiusY).Clamp(0, maxRow);
+                Span<TPixel> sourceRowSpan = sourcePixels.GetRowSpan(offsetY);
+
+                for (int x = 0; x < matrixWidth; x++)
+                {
+                    int offsetX = (column + offsetColumn + x - radiusX).Clamp(offsetColumn, maxColumn);
+                    var currentColor = sourceRowSpan[offsetX].ToVector4();
+                    Vector4Utils.Premultiply(ref currentColor);
```

Comment:
I assume that applying bulk premultiplication outside the convolution might be better, because simpler bulk methods are usually faster. (And it's definitely a better separation of concerns!). 

If any concerns, we should benchmark this!

Reply:
We're only dealing with small numbers but maybe.... I gave it a go outside the loop but soon got lost.

Reply:
Actually forgot that it's all about tiny matrices. They really should be value types (#142), and other optimization attempts are maybe unworthy.

Reply:
`DenseMatrix<T>` is a readonly struct. The indexer is benchmarked to be as fast as raw 1D array indexing.

Reply:
Indexing a 2D struct contains a multiplication which is unnecessary within the inner loop if we have a pointer to the first item of the row.

A note on benchmarking:
In this case, a proper benchmark is a one that iterates through all elements. Using constant index values could be misleading.

But nevermind the whole thing, this is of low importance at this point. I'd rather review the design, than make small improvements on the implementation.

Reply:
Agreed. I'll merge this for now  and we can bash out some designs in the future. I'd like to get this right before we make the processors public.

FYI. Re the benchmark. It benchmarks iterating across entire arrays of different sizes.

https://github.com/SixLabors/ImageSharp/blob/13850d1fec2af49176d08a3576c5e07b7430e476/tests/ImageSharp.Benchmarks/General/Array2D.cs#L113

Reply:
But it has [the unnecessary `count * i` multiplication](https://github.com/SixLabors/ImageSharp/blob/13850d1fec2af49176d08a3576c5e07b7430e476/tests/ImageSharp.Benchmarks/General/Array2D.cs#L70) within the inner loop! 

Also: Going with `Unsafe` and remove the array indexers would strip the benchmarks from the array bounds checking, which is probably the most expensive stuff here, making the results very similar.